### ISSUE TITLE
noConditionalExpect: reports a non-test `assert` function (matches callee by name, ignores the import)

### ISSUE THREAD

**Issue by CarlAllenn** [biomejs/biome#11455]
`nursery/noConditionalExpect` matches the assertion by identifier name and
does not check where the name came from. A local function called `assert`,
unrelated to testing, is reported whenever it is called under a branch.

Biome 2.5.7, macOS arm64, CLI. Repro:
https://biomejs.dev/playground/?lintRules=noConditionalExpect&pane=Diagnostics&language=ts&ruleDomains.test=all&code=aQBtAHAAbwByAHQAIAB7ACAAYQBzAHMAZQByAHQAIAB9ACAAZgByAG8AbQAgACIALgAvAHYAZQByAGIALgBqAHMAIgA7AAoACgBlAHgAcABvAHIAdAAgAGYAdQBuAGMAdABpAG8AbgAgAGQAaQBzAHAAYQB0AGMAaAAoAHMAbAB1AGcAOgAgAHMAdAByAGkAbgBnACwAIABwAGEAeQBsAG8AYQBkADoAIAB1AG4AawBuAG8AdwBuACkAOgAgAHUAbgBrAG4AbwB3AG4AIAB7AAoAIAAgAHMAdwBpAHQAYwBoACAAKABzAGwAdQBnACkAIAB7AAoAIAAgACAAIABjAGEAcwBlACAAIgBhAHMAcwBlAHIAdAAiADoACgAgACAAIAAgACAAIAByAGUAdAB1AHIAbgAgAGEAcwBzAGUAcgB0ACgAcABhAHkAbABvAGEAZAApADsACgAgACAAIAAgAGQAZQBmAGEAdQBsAHQAOgAKACAAIAAgACAAIAAgAHIAZQB0AHUAcgBuACAAbgB1AGwAbAA7AAoAIAAgAH0ACgB9AAoA

```ts
import { assert } from "./verb.js"; // export function assert(p: unknown) { return p; }

export function dispatch(slug: string, payload: unknown): unknown {
  switch (slug) {
    case "assert":
      return assert(payload);
    default:
      return null;
  }
}
```

Reported: `b.ts:6:14 lint/nursery/noConditionalExpect — Unexpected conditional expect() call.`

Control, same run: rename the import to `applyVerb`, change nothing else, and
the finding disappears. So it is the identifier, not the `switch`.

The sibling rule already solves this. `noMisplacedAssertion` documents an
import allowlist (`chai`, `node:assert`, `node:assert/strict`, `bun:test`,
`vitest`, the Deno assert module) and in the same invocation is silent on this
file while `noConditionalExpect` reports it:

| import of `assert` | `noMisplacedAssertion` | `noConditionalExpect` |
| --- | --- | --- |
| `./verb.js` (local) | clean | **reports** |
| `node:assert` | reports | reports |

**Expected:** do not report a call whose callee resolves to an import that is
not an assertion library — the allowlist `noMisplacedAssertion` already uses.

This matters beyond test files: the `test` domain is enabled project-wide
whenever `vitest` is a devDependency, so the rule reaches production modules,
and a test-glob override cannot contain it. The real site was a database write
path.

**Comment by biome-cookie** [comment]
Hello @CarlAllenn, please provide a minimal reproduction. You can use one of the following options:

- Provide a link to [our playground](https://biomejs.dev/playground), if it's applicable.
- Provide a link to GitHub repository. To easily create a reproduction, you can use our interactive CLI via `npm create @biomejs/biome-reproduction`

Issues marked with `S-Needs repro` will be **closed** if they have **no activity within 3 days**.

**Comment by CarlAllenn** [comment]
Reproduction, as requested — both classes in one playground file.

https://biomejs.dev/playground/?lintRules=noConditionalExpect&pane=Diagnostics&language=ts&ruleDomains.test=all&code=aQBtAHAAbwByAHQAIAB7ACAAYQBzAHMAZQByAHQAIAB9ACAAZgByAG8AbQAgACIALgAvAHYAZQByAGIALgBqAHMAIgA7AAoAaQBtAHAAbwByAHQAIAB7ACAAZQB4AHAAZQBjAHQALAAgAHQAZQBzAHQAIAB9ACAAZgByAG8AbQAgACIAdgBpAHQAZQBzAHQAIgA7AAoAZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABkAGkAcwBwAGEAdABjAGgAKABzAGwAdQBnADoAIABzAHQAcgBpAG4AZwAsACAAcABhAHkAbABvAGEAZAA6ACAAdQBuAGsAbgBvAHcAbgApADoAIAB1AG4AawBuAG8AdwBuACAAewAKAHMAdwBpAHQAYwBoACAAKABzAGwAdQBnACkAIAB7AAoAYwBhAHMAZQAgACIAYQBzAHMAZQByAHQAIgA6AAoAcgBlAHQAdQByAG4AIABhAHMAcwBlAHIAdAAoAHAAYQB5AGwAbwBhAGQAKQA7AAoAZABlAGYAYQB1AGwAdAA6AAoAcgBlAHQAdQByAG4AIABuAHUAbABsADsACgB9AAoAfQAKAHQAZQBzAHQAKAAiAGIAbwB0AGgAIABiAHIAYQBuAGMAaABlAHMAIABhAHMAcwBlAHIAdAAiACwAIAAoACkAIAA9AD4AIAB7AAoAdAByAHkAIAB7AAoASgBTAE8ATgAuAHAAYQByAHMAZQAoACIAewB9ACIAKQA7AAoAdABoAHIAbwB3ACAAbgBlAHcAIABFAHIAcgBvAHIAKAAiAHUAbgByAGUAYQBjAGgAYQBiAGwAZQAiACkAOwAKAH0AIABjAGEAdABjAGgAIAAoAGUAcgByAG8AcgApACAAewAKAGUAeABwAGUAYwB0ACgAZQByAHIAbwByACkALgB0AG8AQgBlAEkAbgBzAHQAYQBuAGMAZQBPAGYAKABFAHIAcgBvAHIAKQA7AAoAfQAKAH0AKQA7AA%3D%3D

Two diagnostics: line 6 (a local `assert` imported from `./verb.js`, in a non-test function) and line 16 (a `try`/`catch` where both branches assert).

Class 1 involves an import, so here is the CLI form, measured with Biome 2.5.7 and the two files from the report (`b.ts` importing `assert` from `./verb.js`, config `{ "linter": { "domains": { "test": "all" }, "rules": { "recommended": false, "nursery": { "noConditionalExpect": "on" }, "suspicious": { "noMisplacedAssertion": "on" } } } }`):

```
b.ts:6:14 lint/nursery/noConditionalExpect ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  i Unexpected conditional expect() call.

    4 │   switch (slug) {
    5 │     case "assert":
  > 6 │       return assert(payload);
      │              ^^^^^^^^^^^^^^^
    7 │     default:
    8 │       return null;

c.test.ts:8:5 lint/nursery/noConditionalExpect ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  i Unexpected conditional expect() call.

     6 │     throw new Error("unreachable");
     7 │   } catch (error) {
   > 8 │     expect(error).toBeInstanceOf(Error);
       │     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     9 │   }

Checked 2 files in 1683µs. No fixes applied.
Found 2 infos.
```

Control, same config: renaming the import and its uses to `applyVerb`, changing nothing else, gives `Checked 1 file in 2ms. No fixes applied.` — so the finding is the identifier, not the `switch`. `noMisplacedAssertion` is silent on `b.ts` in the same run.

Expected: a callee resolving to an import that is not an assertion library is not reported (the allowlist `noMisplacedAssertion` already uses), and a `try`/`catch` in which every branch asserts is not reported. Actual: both are reported, the first in production code where no test-glob override can contain it.

**Comment by ematipico** [comment]
Please limit yourself to only one bug report. In this wall of text I can't understand what you're reporting. You reduce this text to minimum and report one bug

**Comment by CarlAllenn** [comment]
Reduced to a single defect: the callee is matched by name, ignoring the import. The try/catch case is dropped.

**Comment by ematipico** [comment]
Aaaand wiped out the issue template in the  process....

**Comment by biome-cookie** [comment]
This issue has been automatically closed because it has been inactive for more than 3 days and it does not have a minimal reproduction.
